### PR TITLE
Add test for correct Nullable setter annotations

### PR DESCRIPTION
**Pull request by wolfs**
This is a follow up to #6920. I added a test using [archunit](https://www.archunit.org/) to check if the setters have the same nullability as the getters. The test currently fails: https://scans.gradle.com/s/te2anb5yj3sqq/tests/nx3baqq2umcn4-3c6c5hjwysoz4

@eskatos I guess we should fix those property accessors, right? When you think this is the right way to implement code constraints, we can go forward and add more checks for even better Kotlin accessibility (like that we need a setter with the same type of the getter) or other architectural constraints on the Gradle code base.

### CODE REVIEW

File: subprojects/architecture-test/src/test/java/org/gradle/architecture/test/KotlinCompatibilityTest.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+            }
+
+            private Set<Accessors> findSymmetricalAccessors(JavaClass javaClass) {
+                return filterGetters(javaClass.getMethods()).stream()
+                    .flatMap(getter -> tryFindMatchingSetters(javaClass, getter).map(s -> new Accessors(getter, s)))
+                    .collect(toSet());
+            }
+
+            private Set<JavaMethod> filterGetters(Set<JavaMethod> methods) {
+                return methods.stream().filter(m -> m.getName().startsWith("get") && m.getName().length() > 3).collect(toSet());
```
Comment (eskatos) [inline subprojects/architecture-test/src/test/java/org/gradle/architecture/test/KotlinCompatibilityTest.java]: :x: this won't find boolean `isName` getters. We should try reuse what we do to e.g. find properties on task types in order to handle all JavaBean spec corner cases.

File: subprojects/architecture-test/src/test/java/org/gradle/architecture/test/KotlinCompatibilityTest.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            }
+
+            private Stream<JavaMethod> tryFindMatchingSetters(JavaClass javaClass, JavaMethod getter) {
+                String setterName = getter.getName().replaceFirst("g", "s");
+                return javaClass.getMethods().stream().filter(m -> m.getName().equals(setterName) && m.getParameters().size() == 1);
```
Comment (eskatos) [inline subprojects/architecture-test/src/test/java/org/gradle/architecture/test/KotlinCompatibilityTest.java]: :x: this should select the setter with the exact same type as the getter

Comment (wolfs) [reply subprojects/architecture-test/src/test/java/org/gradle/architecture/test/KotlinCompatibilityTest.java]: For Kotlin that is enough. I think all the setters should have the same nullability annotations, though.

Comment (eskatos) [reply subprojects/architecture-test/src/test/java/org/gradle/architecture/test/KotlinCompatibilityTest.java]: Sounds good 👍

File: subprojects/architecture-test/src/test/java/org/gradle/architecture/test/KotlinCompatibilityTest.java (hunk 1 above)
Comment (wolfs) [reply subprojects/architecture-test/src/test/java/org/gradle/architecture/test/KotlinCompatibilityTest.java]: I am re-using some of the infrastructure now. I am not very happy about the usage, since I cannot use it fully without adding some extra dependencies to the architecture tests. That is due to the fact that we have some `compileOnly` dependencies like e.g. Zinc and cannot load the classes if those are not on the classpath.

Do you think it is good as it is now or shall I add the dependencies to the classpath?

Comment (eskatos) [reply subprojects/architecture-test/src/test/java/org/gradle/architecture/test/KotlinCompatibilityTest.java]: Got it. It's already much better like that 👍

**Comment by eskatos** [comment]
@wolfs, I'm not ArchUnit api fluent and the javadoc didn't help me much. I'm wondering about properties whose get/setters are spread over a type hierarchy, with overrides and overloads. It would be good to have tests for these cases, I suppose this can be done as a subsequent step though.

**Comment by codecholeric** [comment]
Originally I thought about releasing infrastructure to test the architecture rules as well, but then I refrained because I didn't want to keep that API (which I needed for my own tests) stable. You might be able to grab some code and infrastructure from https://github.com/TNG/ArchUnit/blob/master/archunit-integration-test/src/test/java/com/tngtech/archunit/integration/ExamplesIntegrationTest.java, since that does pretty much what you're talking about I think (execute rules and validate that the expected violations occur on certain examples).
The whole thing might be overkill though, since this tests against JUnit 4 and JUnit 5, etc. But maybe you can reuse parts.

**Comment by wolfs** [comment]
@eskatos This is ready for another round. I now get all the getters/setters from the current class and try to match them up with stuff from the superclass, if I can't find a getter/setter pair in the current class.

Checking the public API is now separate from checking the internal API.

This is the current state: https://e.grdev.net/s/tk5lbdngqscqg/tests

File: subprojects/architecture-test/src/test/java/org/gradle/architecture/test/KotlinCompatibilityTest.java [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+            && (accessorType == PropertyAccessorType.GET_GETTER || accessorType == PropertyAccessorType.IS_GETTER)
+            && m.getParameters().size() == 0
+            && (accessorType != PropertyAccessorType.IS_GETTER || m.getReturnType().isEquivalentTo(Boolean.TYPE) || m.getReturnType().isEquivalentTo(Boolean.class));
+    }
+
+    private static boolean isSetter(JavaMethod m, PropertyAccessorType accessorType) {
+        return !m.getModifiers().contains(JavaModifier.STATIC)
+            && accessorType == PropertyAccessorType.SETTER
+            && m.getParameters().size() == 1;
+    }
```
Comment (eskatos) [inline subprojects/architecture-test/src/test/java/org/gradle/architecture/test/KotlinCompatibilityTest.java]: :o: this will include methods with a non-void return type, I suppose it's ok

File: subprojects/architecture-test/src/test/java/org/gradle/architecture/test/KotlinCompatibilityTest.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+                    return new Accessors(owningClass, propertyName, superclassGetters, setters);
+                }
+            } else {
+                Set<JavaMethod> superclassSetters = superclassAccessors.stream().filter(Accessor::isSetter).map(Accessor::getMethod).collect(toSet());
+                if (!superclassSetters.isEmpty()) {
+                    return new Accessors(owningClass, propertyName, getters, superclassSetters);
+                }
+            }
+            return null;
+        }
```
Comment (eskatos) [inline subprojects/architecture-test/src/test/java/org/gradle/architecture/test/KotlinCompatibilityTest.java]: :o: I would split this method into several explaining methods to make it more clear what happens

File: subprojects/architecture-test/src/test/java/org/gradle/architecture/test/KotlinCompatibilityTest.java (hunk 3 above)
Comment (wolfs) [reply subprojects/architecture-test/src/test/java/org/gradle/architecture/test/KotlinCompatibilityTest.java]: That is our current logic to find setters for properties. I can also check for the void return type. WDYT?

Comment (eskatos) [reply subprojects/architecture-test/src/test/java/org/gradle/architecture/test/KotlinCompatibilityTest.java]: I'd say it's fine for this check to be a bit greedy and also consider setters that return something ; even if they are not javabean compliant at least the consistency of their nullability will be enforced. We can still make it more strict later if it proves to be problematic.

**Comment by wolfs** [comment]
@eskatos I am now ignoring the current violations and `sanityCheck` also runs the architecture tests now. PTAL.
